experimental/GModule/src/GModule.jl

```diff
@@ -1760,6 +1768,144 @@ function hom_base(C::_T, D::_T) where _T <: GModule{<:Any, <:AbstractAlgebra.FPM
+    @show :is_irr
+    return [M]
+  end
+
+  S = []
+  for b = lll_basis(Z_M)
+    z = split_via_endo(b.elem_in_algebra, M)
+    if length(z) > 0
+      append!(S, z)
+    end
+  end
+  return S
+end
+
+function split_homogenous2(M::GModule{<:Any, <:AbstractAlgebra.FPModule{QQFieldElem}})
+  #Steel, p36: SplitHomogeneousByMinimalField(M)
+  #            ... up to Step 2
+  #TODO:-use this to reduce to the m=1 part - ignoring the Schur stuff
+  #      (don't know if possible: rho_F is abs. irr, but as field is too large,
+  #      the restriction of scalars has multiplicity again)
+  #     -use the Amitsur paper instead of Fieker to reduce field? (Does not need
+  #      normality, possibly)
+  #TODO:-does e in Step 2 exists for A-modules (as opposed to G-modules)? Tommy
+  #      spontaneously said yes...
+  #XXX:  Step 3 exists - but requires the field to be normal, hence sucks in general
+  #      the search for e need to be more intelligent and possibly try  more elements
+  #      to find small normal closure
+  #     -or try smaller degree to get some split? Any reduction in m helps
+  #     -use characters to see what we do not want to split: the same module
+  #      might be in a mult 8 or mult 2 component...
+  #     -also do the conic case (for ms = 2)
+  #     -for m=1, s=3: Jessica Cologna reduces to conics as well
+  #
+  #TODO:-make sure the interaction between Hecke (Tommy) and Oscar (Claus)
+  #      is efficient and uses appropriate caching...
+  #     -finally get some examples going and see where the infrastructure can be 
+  #      improved
+  #     - e.g. write & use factored_minpoly, factored_charpoly ...
+  E = endo(M)
+  Z_M = maximal_order(E)
+  @show s = schur_index(E)
+  @show k = dim(center(E)[1])
+  @show m = div(root(div(dim(E), k), 2), s)
+  if m == 1
+    @show :is_irr
+    return [M]
+  end
+
+  for i=basis(E)
+    f = minpoly(i)
+    lf = factor(f)
+    if length(lf) == 1 && degree(f) == s*m*k &&  lf[f] == 1
+      @show :bingo
+      Ka, a = number_field(f)
+      Mf = extension_of_scalars(M, Ka)
+      h = hom(Mf, Mf, hom(Mf.M, Mf.M, map_entries(Ka, i.matrix) - a*identity_matrix(Ka, dim(M))))
+      k = kernel(h)
+      #k is abs. irr. but field is too large
+      return k
+    end
+  end
+
+  return :nothing_found
+end
+
+function split_into_homogenous(M::GModule{<:Any, <:AbstractAlgebra.FPModule{QQFieldElem}})
+  #Steel, p28: HomogeneousComponents(M)
+  #Careful: the list in the end contains homogenous components - but
+  #         with lots of repetition
+  #TODO: write and use CentreOfEndomorphismRing
+  #      have a sane overall strategy
+  E = endo(M)
+  C, mC = center(E)
+  H = []
+  for b = basis(C)
+    f = minpoly(b)
+    lf = factor(f)
+    if length(lf) == 1
+      if degree(f) == dim(C)
+        return [M]
+      end
+      continue
+    end
+    for (p, k) = lf
+      x = (p^k)(b)
+      h = hom(M, M, hom(M.M, M.M, matrix(mC(x))))
+      append!(H, split_into_homogenous(kernel(h)[1]))
```

Comment:
Concerning duplicate submodules:
My understanding is that once we have found one element in `basis(C)` whose minimal polynomial has two different factors, `M` is the direct sum of kernel and image of what is called `x` here.
Thus we can call `split_into_homogenous` for kernel and image, and return the concatenation of the results.